ports/nrf/supervisor/port.c

```diff
@@ -247,7 +248,15 @@ void port_sleep_until_interrupt(void) {
         sd_app_evt_wait();
     } else {
         // Call wait for interrupt ourselves if the SD isn't enabled.
+        // Note that `wfi` should be called with interrupts disabled,
+        // to ensure that the queue is properly drained.  The `wfi`
+        // instruction will returned as long as an interrupt is
+        // available, even though the actual handler won't fire until
+        // we re-enable interrupts.
+        common_hal_mcu_disable_interrupts();
+        RUN_BACKGROUND_TASKS; // Drain any buffers remaining
```

Comment:
background task are run in the caller `mp_hal_delay_ms()`, maybe we don't have to run it here, but rather move/sort it out. I am not so sure though.

Reply:
Correct, but you must run background tests after you disable interrupts, to ensure that e.g. the USB queues are empty.

Reply:
yeah, I know that intention of yours, I am thinking a way to make tinyusb maybe cooperate better with wfi() :thinking:  :thinking:

Reply:
@tannewt  @xobs we shouldn't run background task with interrupt disabled. This particular issue is mentioned by TockOS (note 3) a few months ago https://github.com/hathach/tinyusb/issues/279 . And I overlooked it. The correct implementation should be an function to check if there is anything to run with the event queue of tud_task(), if yes, don't enter WFI(), re-enable interrupt, and loop over again.

Reply:
There are two approaches that we can take, and I think @tannewt should decide which approach they think is best:

1. Allow background tasks to be run with interrupts disabled. This is the approach that's started by this patch, but there may be many other minefields in other code that requires interrupts to be enabled in order to function properly
2. Add a new `SHOULD_RUN_BACKGROUND_TASKS` macro that polls each subsystem, similar to how `RUN_BACKGROUND_TASKS` runs each subsystem. This `SHOULD_RUN_BACKGROUND_TASKS` will include a call to `tud_task_is_queue_empty()`, among other calls.

Am I correct in stating you're advocating for approach 2? This patch takes approach 1, because I thought it would be easier to implement and require less code overall. Also because `tud_task_is_queue_empty()` didn't exist. Approach 2 does benefit from potentially allowing non-critical-section-safe code from going into the `RUN_BACKGROUND_TASKS` macro, which seems cleaner.

Reply:
yeah you are spot-on and I am adding the usb event queue count API() just now. Your PR has already implemented 99% of the work. We only need to change the `run background` to `if still has more background then skip` before executing wfi().

Reply:
PR for tinyusb https://github.com/hathach/tinyusb/pull/414 is on the way.

Reply:
@xobs I was thinking 2 would be our approach to background tasks long term but @jepler prototyped a callback queue in #2879